float32[] body_velocity